core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -232,6 +305,74 @@ public boolean isServing() {
     return configInfoList;
   }
 
+  @Override
+  public long getMasterId(String hostname) {
+    MetaMasterInfo existingMaster = mMasters.getFirstByField(HOSTNAME_INDEX, hostname);
+    if (existingMaster != null) {
+      // This master hostname is already mapped to a master id.
+      long oldMasterId = existingMaster.getId();
+      LOG.warn("The master {} already exists as id {}.", hostname, oldMasterId);
+      return oldMasterId;
+    }
+
+    MetaMasterInfo lostMaster = mLostMasters.getFirstByField(HOSTNAME_INDEX, hostname);
+    if (lostMaster != null) {
+      // This is one of the lost masters
+      synchronized (lostMaster) {
+        final long lostMasterId = lostMaster.getId();
+        LOG.warn("A lost master {} has requested its old id {}.", hostname, lostMasterId);
+
+        // Update the timestamp of the master before it is considered an active master.
+        lostMaster.updateLastUpdatedTimeMs();
+        mMasters.add(lostMaster);
+        mLostMasters.remove(lostMaster);
+        mMasterConfigChecker.lostNodeFound(lostMaster.getId());
+        return lostMasterId;
+      }
+    }
+
+    // Generate a new master id.
+    long masterId = IdUtils.getRandomNonNegativeLong();
+    while (!mMasters.add(new MetaMasterInfo(masterId, hostname))) {
+      masterId = IdUtils.getRandomNonNegativeLong();
+    }
+
+    LOG.info("getMasterId(): Hostname: {} id: {}", hostname, masterId);
+    return masterId;
+  }
+
+  @Override
+  public void masterRegister(long masterId, RegisterMasterTOptions options)
+      throws NoMasterException {
+    MetaMasterInfo master = mMasters.getFirstByField(ID_INDEX, masterId);
+    if (master == null) {
+      throw new NoMasterException(ExceptionMessage.NO_MASTER_FOUND.getMessage(masterId));
+    }
+    synchronized (master) {
+      master.updateLastUpdatedTimeMs();
+    }
+
+    List<ConfigProperty> configList = options.getConfigList().stream()
```

Comment:
Will an option always have a configList?

Reply:
For master communication, we do not need to support backward compatibility. If the standby masters are in old versions, it will not register with the leader master. If it is in the new version, it will always register with the leader master with configuration.